### PR TITLE
Allow command extensions to use a custom base class. 

### PR DESCRIPTION
**Pull request by phurley**
Tested using a java project, and the generated code looks acceptable in C++, but not tested
on an actual robot. Similar changes are probably warranted for other types of extensions, if there is a ClassName field, it should be used.

FYI thanks for robot builder, it is a huge help getting students started and providing a very reasonable structure for their robot code.

### CODE REVIEW

**Comment by SamCarlberg** [comment]
Can you give an example extension that uses this?

**Comment by phurley** [comment]
I have a simple subclass of Command, called TimedCommand. In the properties, I include an option to set the timeout, and the idea is that it will automatically handle the isFinished with the correct timeout. We would also like to have the option of using a team specific base class, which adds some simple logging to commands as well.

**Comment by SamCarlberg** [comment]
If you want commands to automatically timeout, you don't need a custom subclass. Just use [this constructor](http://first.wpi.edu/FRC/roborio/release/docs/java/edu/wpi/first/wpilibj/command/Command.html#Command-java.lang.String-double-) instead of the one that RobotBuilder automatically uses.

For the logging commands, if there a reason why adding logger calls or print statements where you need them won't work?

**Comment by phurley** [comment]
The code could be written, by hand, in both cases. In the first case, the TimedCommand would provide the isFinished method, so they would not have to write the same code multiple times. In the case of the logging, again they could write the log statements, but it would be nice if the parent class could provide that functionality without having to repeat it every time.

Again all of these operations could be done by repeating code in each command, the idea would be to allow for command operations to be pushed down to a subclass of Command. If this is not the direction you want to go, no worries -- but you might then want to at least make a comment in the screen step describing extensions that it is not supported, because based on the current documentation, I expected this to be supported.

**Comment by phurley** [comment]
We have a similar issue with a subsystem. We have a LoopingSubsystem class, which inherits from Subsystem and provides a loop method that uses your Notifier to call it in the background. Ideally, it could be instantiated inside robot builder, generating the correct extends line (even better with a custom constructor/loop method already in place, but those are easily added if needed).

Without robot builder support, the team has to either repeat the boilerplate or add the functionality into a member class, and write a bunch of "pass through" methods to expose the feature to Commands.

I would be happy to add support for this as well, but don't want to work through it, if this is not the direction you want to pursue.

**Comment by sciencewhiz** [comment]
@SamCarlberg using that constructor only works if you also check isTimedOut in isFinished, it isn't all done automatically. 

I'll try this out. We also extended Command to add logging and a slightly different state sequence. We modified robotbuilder to generate code using our command, but this should allow us to get by with just extensions.

**Comment by sciencewhiz** [comment]
@phurley Do you have an example extension for one of your commands that works with this PR? I tried to make one but can't get Robot Builder to recognize it.

**Comment by phurley** [comment]
Sure, find it attached. FYI I was also thinking that the right fix here would be to include the super.initialize(), return super.isFinished(), etc in the default template, so minimize surprises for teams.

[TimedCommand.zip](https://github.com/wpilibsuite/RobotBuilder/files/616071/TimedCommand.zip)

File: src/main/resources/export/cpp/Command.h [hunk 1]
```diff
@@ -13,7 +14,7 @@
  *
  * @author ExampleAuthor
  */
-class #class($command.name): public Command {
+class #class($command.name): public #klass($command) {
```
Comment (sciencewhiz) [inline src/main/resources/export/cpp/Command.h]: Need the same type of change in Command-constructor.cpp

**Pull request merged**